### DISCUSSION TITLE
Understanding smartnoise SQL `.execute` and `get_privacy_cost`

### DISCUSSION THREAD

**Discussion by huwd** [opendp/opendp#379] [SmartNoise] [answered]
Hello, 

We've been experimenting with the [smartnoise.sql package](https://docs.smartnoise.org/en/latest/api-reference/opendp.smartnoise.sql.html).

We've been trying to figure out some of the design intent behind the `.execute` command.

Specifically, if I had made a `Privacy` object, say with an epsilon of `0.1`, and a Delta of `1/{rough count of individuals in our data set}`. And then I wrap that around a large dataset.

I'd expected that if I start doing `.execute` with SQL that would produce an overly specific result, I would see some kind of failure senario where it cannot guarentee a 0.1 epsilon and warns me? Instead we get back data, but it is difficult to tell what the accuracy might be?

I had hoped I could use `.get_privacy_cost` to perform checks before `.execute`. Perhaps with the intention of warning the user "your query cannot guarentee privacy to the required epsilon". But instead I was suprised to see that pretty much any query returned the tuple, the Delta would vary, but then seemed to platau at a particularly low decimal value.

Q: Is there a way of using `.execute` and `get_privacy_cost` to advise a user that the querys they are writing are incapable of meeting the passed epsilon values?

I'd hoped I could define my desired epsilon in advance, then guarentee some bounds of accuracy or probability of compromsie with delta, and get smartnoise.sql to error or provide feedback if those conditions were not met?

I suspect the gap is me missing something more fundimental about smartnoise or the parameters here?

**Comment by joshua-oss** [comment, marked as answer]
Hello,

Thanks for raising this topic.  The SQL layer spends budget in increments of the epsilon set in the `Privacy` object, as described here: https://github.com/opendp/smartnoise-sdk/blob/main/sql/docs/papers/DP_SQL_budget.pdf

So, for example, a `SELECT SUM(age)` would use 1X the epsilon, and `SELECT SUM(age), SUM(income)` would use (roughly) 2X the epsilon.  A `SELECT AVG(age)` would spend ~2X the epsilon, because we compute average from a SUM and a COUNT, both of which require noisy mechanism invocation.  On the other hand, a `SELECT AVG(age), COUNT(age), SUM(age)` would only result in a ~2X epsilon spend, because the second and third column can be re-used to compute the first column without additional noise.  Also note that some statistics, such as COUNT and SUM, are (epsilon, 0) private, while we need to use (epsilon, delta) where counts are used to censor rare dimensions.  A single query will have at most 1 invocation of the dimension-censoring mechanism, so the per-query delta will be stable regardless of the number of columns.  However, delta over the life of a session will compose.

We keep track of total privacy cost in the `odometer.spent` property on the private connection.  This is the total spend for all queries that have been issued since the connection was created.  The odometer spent can be substantially lower than what would be obtained by summing the cost of each column or query, both because of tighter composition, and because of potential for column re-use.  Therefore, checking the odometer is the best way to know what total spend is.

Regarding rationale, we assume that a session will include many different queries against the database, with the workload unknown ahead of time.  And we want the privacy cost and accuracy of two queries that aggregate different columns to be consistent with the cost of a single query that aggregates both columns.  IOW, we use a fixed epsilon per released statistic, rather than a fixed epsilon per query or batch, to ensure that noise scales don't change in unpredictable ways depending on how the analyst choose to concatenate columns.

And, because the composition rules allow substantially smaller cost than the sum of costs, it's not possible to know the total spend for a session without looking at the entire batch of queries.  IOW, extrapolating the total cost of a session by multiplying the number of queries by the result of `get_query_cost`, would give poor results.  The results of `get_query_cost` are accurate if that query is the only query in the session, but are mainly useful for getting a sense of what epsilon multiple the query is using (e.g. to see that a count and sum are being re-used to preserve budget).  For a single query, the composition will typically devolve to basic composition, which is pessimistic.

I think we could implement something like what you're describing, though.  For example, we could make a version of `get_query_cost` that takes in an entire batch of queries, and computes total cost for a fixed noise scale, or alternately a related method that takes a fixed budget cost and determines what noise scale would be needed to implement the whole batch within the budget constraints.  Both require the analyst to know the whole query workload in advance.  Can you describe a bit more about the workflow for your ideal scenario?

**Reply by huwd** [reply]
Thanks so much, that's really helpful

**Reply by huwd** [reply]
I've had a bit more time to go over this, happy to feed back on your last question.

Your answer has really helped - I think the scenario you are describing may not be relevant for us anyway.
Our notebook is importng a large volume of log based data, performing a bit of pre-processing to concatentae a few values and produce a unique ID (I suspect we could do this within a PrivacyReader but I'll look int that later).

Then we're wrapping the result in a PrivacyReader and firing off a fairly simple SQL statmenet with a count of values, grouped_by day with the aim of producing a differentially private aggegate.

So our "session" if I've understood correctly only actually extends over a single differentially private `.execute` query.
We are doing a couple of things there like a "where" clause that disgards some irrelevant data. I'm a bit on the fence about if we should be doing that as part of any "pre-processing" or if we'll get a more private / appropraitely accurate result if we do that in the SQL we pass to `.execute` 🤔 .

But in any case, the intention in this scenario is to output a single differenentially private set of daily aggregates.

If an analyist wanted to drill down there further, my advice was to go back and extend that notebook, so we make the where clause more specific (perhaps limiting it to one area of our logs before we run the aggregates).

In that way we should just be looking at a single query session for the time being.

The batch query feature you describe is interesting, but I can see how you'd need to collect all the SQL in advance.
I suppose a question I'd ask the analyist is:

Why could you not go back and write this all into a single more specific SQL query to output your result?

I'm generally thinking we get better results if we go:
Pre-processed data set => Smartnoise.sql.execute(sql_statement) => Aggregate to take away

Though i know there are some ways of outputting a differentially private dataset for further analysis 🤔, seeking aggregates seems simpler as we're still learning the basic concepts right now?

**Reply by joshua-oss** [reply]
Thanks for following up with more about your scenario.  It sounds like you're on the right track.

Doing a lot of the work in preprocessing can make sense for a variety of reasons:
* The analyst is more comfortable/productive with a different data processing approach, such as using dplyr in R
* The pre-processing needs to use some SQL that SmartNoise doesn't know how to validate.  For example, the source data might be normalized relation tables with primary-foreign key relationships.  Creating a view to expose the data in SmartNoise's expected format would be done before calling SmartNoise
* Feature engineering, such as log transform or normalization, creating indicator variables, etc.

And so on.  Your intuition is correct, that preprocessing (like postprocessing) doesn't incur any privacy cost.  The privacy cost comes when noise is applied. The caveat with preprocessing is that the analyst needs to be sure that the result of pre-processing is consistent with the upper and lower bounds in the metadata, enforces the max_ids specified, etc.

This latter point can be tricky.  Preprocessing shouldn't use the "global" picture of the data for any operations.  For example, if the preprocessing looks at the data to determine min and max, and then uses min-max normalization to normalize the column to 0.0-1.0, that would leak privacy.  If the min and max are public knowledge, such as when ages are arbitrarily clamped to (0, 100), this isn't an issue.  The problem is when the preprocessing step looks at the data to determine what to do.  Same with normalization that uses the sum or mean.  And feature engineering that is trained on the data, such as k-means centroids or embedding or etc could also leak privacy.  If you can find a transform based off of public data or provided by a domain expert, it's OK; just can't look at the global picture to make a per-row transform.

Filtering is a special case, because you don't need to compose privacy cost over queries that operate over disjoint sets of users.  For example, if the data has a 'gender' column, and each user can only have one value, filtering to gender==0 and gender==1 would always operate on different sets of users.  In this case, if you did the filtering in preprocessing and issued two different queries with epsilon=0.5, the total spend would look like 1.0.  But if you specified the query as "GROUP BY gender", the API will know that these are disjoint, and would charge only 0.5.  Doing the filtering steps through preprocessing is fine, but you'd want to make sure you don't overcharge yourself for queries over disjoint partitions.

Note that SmartNoise isn't currently smart enough to realize that two sequential queries over disjoint keys would be disjoint.  For example, if you do a "GROUP BY gender", it will return results for each gender and only charge you once.  But if you do a separate query for each gender, it will charge you multiple times.

**Reply by Nyzl** [reply]
Hi,

I'm working with Huw on this use of DP and wanted to provide some context around expected outputs.

My ideal process is to take a private table that contains:
date, url, client_id
and generate a differentially private table of aggregates:
date, url, count distinct of client_id

It should be noted that there is a long tail of urls with single numbers of client_ids, and a lot with 1 id, and that a client_id could exist in multiple rows.

The initial raw table has 2.5m rows which is reduced to 315k rows when doing the distinct count, but I just can not get a private output near that row count.
I've tried varying the epsilon and delta and only a delta of 1 gets close. So now I'm wondering if there is something in the yml file.

Database:
  cdn:
    cdn:
      row_privacy: False
      rows: 2509848
      max_ids: 100
      sample_max_ids: True
      cdn_date:
        type: string
      url:
        type: string
      client_id:
        type: string
        private_id: True

And at the moment it seems like whenever I use the max_ids, the number of rows output falls dramatically.

There is a clear gap in my knowledge with this subject.
Any pointers would be gratefully received

**Comment by joshua-oss** [comment]
Hi @Nyzl 

This looks like a case of unbounded user contribution.  This is usually a challenge with data that have a zipf-like distribution.  There are two issues: 1) sensitivity could be very high, and 2) we need to censor infrequent dimensions.

1) Sensitivity: Differential privacy ensures that the presence or absence of any single individual will not be distinguishable from the output.  To do this, the algorithm needs to know how much any individual can conceivably impact the output.  This is the "sensitivity".  For example, if each user appears only once, and the quantity is SUM(age), and assuming ages are truncated to 100, each user could potentially affect the output by up to 100, so the sensitivity would be 100.  A larger sensitivity means more noise is added.

If each user can contribute only one row, and the statistic being computed is a COUNT, the sensitivity will be 1, and noise will be relatively small.

The baseline setting in many differential privacy papers is that each user contributes only one row, so the sensitivity is only based on the range of the value.  In datasets where the user can supply more than one row, we multiply the sensitivity by the number of rows the user can contribute.  For example, if each user can contribute 10 rows, and the quantity is SUM(age), sensitivity would be 1000.

With `sample_max_ids` set to true, we will automatically sample to ensure that the user contribution remains bounded to the number of rows specified in `max_ids`.  This will ensure that there are no privacy leaks, but could still result in excess protection.  For example, if the maximum number of record per user is 12, but median is 3, setting `max_ids` to 10 will be safe, but most users will be getting a lot of extra noise, just to protect the users who have 10 or more.  This is because we need to provide full protection for the worst-case scenario (i.e. a user with 10 rows).  So, you normally want to set the user contribution as low as possible.

It is even more challenging in scenarios where each user can appear in every bucket of the GROUP BY clause.  For example, `SELECT url, COUNT(DISTINCT user) FROM click_log` or `SELECT ngram, COUNT(DISTINCT user) FROM email_corpus`.  In these cases, the user contribution is unbounded, since each user could theoretically have clicked on every URL.  So, if there are a billion URLs, a user could theoretically contribute a billion rows, and sensitivity would be 1 billion.  Sampling is essential in this case, as you are doing by setting `max_ids` to 100.  However, 100 is still a pretty large noise scale for counts.

One workaround is to reduce the number of rows that each user can contribute (e.g. set `max_ids` to 1 or 2).  This lowers the sensitivity, leading to more accurate counts.  However, this might introduce some bias, by undercounting URLs that are disproportionately visited by the most active users.  You can try a few queries with this sort of sampling and no noise added to the counts, to see if the bias is an issue, and to determine what setting of `max_ids` reaches a suitable lack of bias.  Another workaround would be to simply sample the rows uniformly, which will avoid bias, but could lose some accuracy.  In your example, for `max_ids` of 100, you could sample uniformly at 1% and set `max_ids` to 1.

None of these are particularly great solutions, since they either risk introducing some bias, or they provide more protection than the vast majority of individuals will need.

2) Censoring infrequent dimensions.  With something like URLs or ngrams, the number of dimensions is essentially unbounded as well, which risks uniquely identifying a user.  For example, image someone visits a URL like "email.BobbyUniqueSurname.com".  COUNT DISTINCT will result in a count of 1, and even if the noisy count ends up being something like -2, the fact that the URL even shows up in the output will reveal that Bobby was checking his email.  To be differentially private, we need to randomize the counts, then drop dimensions that fall below a certain threshold.  This works well when the number of users is uniformly distributed, but creates a ton of churn in the long tail where counts are very small.

We have another way to choose the dimensions, which is more efficient with tail-distributed data.  You can see a sample of calling it here: https://github.com/opendp/smartnoise-sdk/blob/main/sql/samples/Policy%20Laplace%20on%20Spark.ipynb

The algorithm does not return counts, but it tells which URLs are safe to keep for a given epsilon.  Then you could make a second pass where you do count distinct for only those URLs, and turn off the `censor_dims` in metadata (since they've already been censored).  This would disable the extra noise used for censoring and result in more accurate counts.

Which of these three approaches to use would depend on your needs.  If you don't have a lot of data, but can tolerate some bias, dropping max_ids to 1 would probably be best.  If you have lots of data, such that sampling down to 1% (e.g.) wouldn't much affect accuracy, I'd go with that.  If the scenario is somewhere in-between, you can try some different mix of those two approaches to see what the best settings are.  Finally, if you need to compute the rankings regularly (such as daily), but the list of "safe" URLs to report will be changing less frequently (such as monthly), I would use the "Policy Laplace" link above to get a monthly list of safe URLs, and then use those URLs in the daily processing.

**Reply by Nyzl** [reply]
Thank you so much @joshua-oss, there is a lot to unpick and digest there which will really help our understanding of whats going on under the hood and help us apply DP to our problem
